File: julia/src/name.jl
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

abstract type AbstractNameManager end
const NameType = Union{Base.Symbol, AbstractString}
const NameCounter = Dict{Base.Symbol, Int}

import Base: get!

# Default implementation for generating a name for a symbol.
# When a name is specified by the user, it will be used. Otherwise, a name
# is automatically generated based on the hint string.
function _default_get_name!(counter :: NameCounter, name :: NameType, hint :: NameType)
  if isa(name, Base.Symbol) || !isempty(name)
    return Symbol(name)
  end

  hint = Symbol(hint)
  if !haskey(counter, hint)
    counter[hint] = 0
  end
  name = Symbol("$hint$(counter[hint])")
  counter[hint] += 1
  return name
end

mutable struct BasicNameManager <: AbstractNameManager
  counter :: NameCounter
end
BasicNameManager() = BasicNameManager(NameCounter())

function get!(manager :: BasicNameManager, name :: NameType, hint :: NameType)
  _default_get_name!(manager.counter, name, hint)
end

mutable struct PrefixNameManager <: AbstractNameManager
  prefix  :: Base.Symbol
  counter :: NameCounter
end
PrefixNameManager(prefix :: NameType) = PrefixNameManager(Symbol(prefix), NameCounter())

function get!(manager :: PrefixNameManager, name :: NameType, hint :: NameType)
  name = _default_get_name!(manager.counter, name, hint)
  return Symbol("$(manager.prefix)$name")
end

DEFAULT_NAME_MANAGER = BasicNameManager()

File: csharp-package/src/MxNet/Name.cs
/*****************************************************************************
   Copyright 2018 The MxNet.Sharp Authors. All Rights Reserved.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
******************************************************************************/
using System.Collections.Generic;
using System.Threading;

namespace MxNet
{
    public class Name
    {
        public class NameManager : MxDisposable
        {
            internal static ThreadLocal<NameManager> current = new ThreadLocal<NameManager>();

            private readonly Dictionary<string, int> counter = new Dictionary<string, int>();
            private NameManager old_manager;

            public static NameManager Current
            {
                get
                {
                    if (current.IsValueCreated)
                        return current.Value;

                    return new NameManager();
                }
                set => current.Value = value;
            }

            public override MxDisposable With()
            {
                old_manager = Current;
                current.Value = this;
                return this;
            }

            public override void Exit()
            {
                if (old_manager != null) current.Value = old_manager;
            }

            public virtual string Get(string name, string hint)
            {
                if (!string.IsNullOrWhiteSpace(name)) return name;

                if (!counter.ContainsKey(hint)) counter[hint] = 0;

                name = hint + counter[hint];
                counter[hint]++;
                return name;
            }
        }

        public class Prefix : NameManager
        {
            private readonly string _prefix;

            public Prefix(string prefix)
            {
                _prefix = prefix;
            }

            public override string Get(string name, string hint)
            {
                var r = base.Get(name, hint);
                return _prefix + r;
            }
        }
    }
}